include/RAJA/policy/sycl/kernel/For.hpp

```diff
@@ -0,0 +1,478 @@
+          typename Types>
+struct SyclStatementExecutor<
+    Data,
+    statement::For<ArgumentId, RAJA::sycl_local_123_direct<Dim>, EnclosedStmts...>,
+    Types> {
+
+  using stmt_list_t = StatementList<EnclosedStmts...>;
+
+  // Set the argument type for this loop
+  using NewTypes = setSegmentTypeFromData<Types, ArgumentId, Data>;
+
+  using enclosed_stmts_t =
+      SyclStatementListExecutor<Data, stmt_list_t, NewTypes>;
+
+
+  static
+  inline RAJA_DEVICE void exec(Data &data, cl::sycl::nd_item<3> item, bool thread_active)
+  {
+    auto len = segment_length<ArgumentId>(data);
+    auto i = item.get_local_id(Dim);
+
+    // assign thread id directly to offset
+    data.template assign_offset<ArgumentId>(i);
+
+    // execute enclosed statements if in bounds
+    enclosed_stmts_t::exec(data, item, thread_active && (i<len));
+
+  }
+
+  static
+  inline
+  LaunchDims calculateDimensions(Data const &data)
+  {
+    auto len = segment_length<ArgumentId>(data);
+
+    // request one block per element in the segment
+    LaunchDims dims;
+    if (Dim == 0) {
+      dims.local.x = len;
+    }
+    if (Dim == 1) {
+      dims.local.y = len;
+    }
+    if (Dim == 2) {
+      dims.local.z = len;
+    }
+
+    // combine with enclosed statements
+    LaunchDims enclosed_dims = enclosed_stmts_t::calculateDimensions(data);
+    return dims.max(enclosed_dims);
+  }
+};
+
+/*
+ * Executor for local item sharing loop inside SyclKernel.
+ * Provides a local-stride loop (stride of work item local range)
+ * for each item in dim.
+ * Assigns the loop index to offset ArgumentId
+ */
+template <typename Data,
+          camp::idx_t ArgumentId,
+          int Dim,
+          typename... EnclosedStmts,
+          typename Types>
+struct SyclStatementExecutor<
+    Data,
+    statement::For<ArgumentId, RAJA::sycl_local_123_loop<Dim>, EnclosedStmts...>,
+    Types> {
+
+  using stmt_list_t = StatementList<EnclosedStmts...>;
+
+  // Set the argument type for this loop
+  using NewTypes = setSegmentTypeFromData<Types, ArgumentId, Data>;
+
+  using enclosed_stmts_t =
+      SyclStatementListExecutor<Data, stmt_list_t, NewTypes>;
+
+
+  static
+  inline RAJA_DEVICE void exec(Data &data, cl::sycl::nd_item<3> item, bool thread_active)
+  {
+    auto len = segment_length<ArgumentId>(data);
+    auto i0 = item.get_local_id(Dim);
+    auto i_stride = item.get_local_range(Dim);
+    auto i = i0;
+
+    for(; i < len;i += i_stride){
+
+      // Assign the x thread to the argument
+      data.template assign_offset<ArgumentId>(i);
+
+      // execute enclosed statements
+      enclosed_stmts_t::exec(data, item, thread_active);
+    }
+    // do we need one more masked iteration?
+    if(i - i0 < len)
+    {
+      // execute enclosed statements one more time, but masking them off
+      // this is because there's at least one thread that isn't masked off
+      // that is still executing the above loop
+      enclosed_stmts_t::exec(data, item, false);
+    }
```

Comment:
I'm pretty sure this was changed in the cuda impl to use a single loop so the threads execute the same line of code at the same time as a way of avoiding divergence.